[SYCL][LLVM] Adding property set I/O library for SYCL

asudarsa opened the pull request:
SYCL objects are normally associated with properties that provide additional information about those objects. This information should be preserved during compilation and made available to the SYCL runtime via property sets. This PR provides an implementation of a simple library which allows to construct and serialize/de-serialize a sequence of typed property sets, where each property is a <name,typed value> pair.

PropertyValue set format:
   '['<PropertyValue set name>']'
   <property name>=<property type>'|'<property value>
   <property name>=<property type>'|'<property value>
   ...
   '['<PropertyValue set name>']'
   <property name>=<property type>'|'<property value>
where
   <PropertyValue set name>, <property name> are strings
   <property type> - string representation of the property type
   <property value> - string representation of the property value.

For example:
[Staff/Ages]
person1=1|20
person2=1|25
[Staff/Experience]
person1=1|1
person2=1|2
person3=1|12

This library will be used in the SYCL offloading compilation flow to pass information about specialization constants, etc, to the SYCL runtime.

asudarsa committed:
[SYCL][LLVM] Adding property set I/O library for SYCL

SYCL objects are normally associated with properties that provide additional information about those objects. This information should be preserved during compilation and made available to the SYCL runtime via property sets.
This PR provides an implementation of a simple library which allows to construct and serialize/de-serialize a sequence of typed property sets, where each property is a <name,typed value> pair.

PropertyValue set format:
   '['<PropertyValue set name>']'
   <property name>=<property type>'|'<property value>
   <property name>=<property type>'|'<property value>
   ...
   '['<PropertyValue set name>']'
   <property name>=<property type>'|'<property value>
where
   <PropertyValue set name>, <property name> are strings
   <property type> - string representation of the property type
   <property value> - string representation of the property value.

For example:
[Staff/Ages]
person1=1|20
person2=1|25
[Staff/Experience]
person1=1|1
person2=1|2
person3=1|12

This library will be used in the SYCL offloading compilation flow to pass information about specialization constants, etc, to the SYCL runtime.

Signed-off-by: Arvind Sudarsanam <arvind.sudarsanam@intel.com>

asudarsa commented:
@jhuber6

Can you please take a look at this PR when convenient? This support is required to add SYCL offloading compilation support.
It is specifically used in the sycl-post-link step.

Thanks

jhuber6 commented on llvm/include/llvm/Support/PropertySetIO.h:
```diff
@@ -0,0 +1,256 @@
+  void copy(const PropertyValue &P);
+
+  Type Ty = NONE;
+  // TODO: replace this union with std::variant when uplifting to C++17
```
LLVM uses C++17, so the time is now I guess.

asudarsa replied:
Good catch. I will update this and look for other sites.
Thanks

asudarsa replied:
Added in 526633fe54188134d77e033c512fafbab11ff4ba

Thanks

jhuber6 commented on llvm/include/llvm/Support/Base64.h:
```diff
@@ -56,6 +56,32 @@ template <class InputBytes> std::string encodeBase64(InputBytes const &Bytes) {
+// General-purpose Base64 encoder/decoder class wrapper.
+class Base64 {
+public:
+  using byte = uint8_t;
```
Can we use `std::byte`?

asudarsa replied:
Thanks for the pointer. Since we have moved to C++17 and we intend only memory accesses here, std::byte is a good match. Will add the change in upcoming patch.

asudarsa replied:
Added in 526633fe54188134d77e033c512fafbab11ff4ba

Thanks

jhuber6 commented on llvm/include/llvm/Support/PropertySetIO.h:
```diff
@@ -0,0 +1,256 @@
+
+  ~PropertyValue() {
+    if ((getType() == BYTE_ARRAY) && Val.ByteArrayVal)
+      delete[] Val.ByteArrayVal;
```
This this need to be a raw pointer?

asudarsa replied:
Yes. Modified it. Thanks

asudarsa replied:
For our implementation, this pointer is always owned by the caller of 'read' and 'write' functions. Hence we do not need to delete it here.

Thanks

asudarsa committed:
Move from union to variant

Signed-off-by: Arvind Sudarsanam <arvind.sudarsanam@intel.com>

jhuber6 reviewed (commented):
Can you give me some background on the problem we're trying to solve here? This seems to be the serialization of some kind of binary metadata. I think the go-to solution for that is stuff like `yaml2obj` or `obj2yaml`. For example, that's what I did for the `OffloadBinary` class and it's what AMDGPU uses for its kernel metadata.

asudarsa committed:
Add missing implementation of encode/decode functions

Signed-off-by: Arvind Sudarsanam <arvind.sudarsanam@intel.com>

asudarsa commented:
> Can you give me some background on the problem we're trying to solve here? This seems to be the serialization of some kind of binary metadata. I think the go-to solution for that is stuff like `yaml2obj` or `obj2yaml`. For example, that's what I did for the `OffloadBinary` class and it's what AMDGPU uses for its kernel metadata.

Hi @jhuber6 

You are right. We are trying to represent a unique configuration of kernel metadata using a specific data structure called PropertySet. We provide encode/decode functionality using Base64 representation so that the metadata can be serialized and written to a string that can be included in the StringMap of the OffloadingImage. The SYCL offload wrapper decodes this string and the decoded property set is included into SYCL binary image. SYCL runtime has been programmed to identify the property sets and extract required information during runtime.
This support was added to our downstream repository way back in 2021 and has provided us with a well-oiled infrastructure to transmit kernel properties across multiple stages of compilation and runtime. 
I briefly looked at yaml support to represent kernel metadata and it does seem like a possible fit for our purposes.

Our current objective is to get an end-to-end compilation+runtime flow for SYCL available in upstream LLVM. Once available, we can surely look at both approaches and try to come up with a unified solution.

Please let me know if this is agreeable.
Thanks again for your feedback thus far.

Sincerely

asudarsa commented on llvm/lib/Support/Base64.cpp:
```diff
@@ -90,3 +90,180 @@ llvm::Error llvm::decodeBase64(llvm::StringRef Input,
   }
   return Error::success();
 }
+
```
Existing encode/decode logic works well for char arrays. We need something that will work for std::byte arrays.

Thanks

jhuber6 commented on llvm/include/llvm/Support/PropertySetIO.h:
```diff
@@ -0,0 +1,259 @@
+//==-- PropertySetIO.h -- models a sequence of property sets and their I/O -==//
```
Is support where this should go? Feels more like an object thing.

asudarsa replied:
I think 'Support' is a better option. It is just yet another data format.

Thanks

arsenm commented on llvm/unittests/Support/PropertySetIOTest.cpp:
```diff
@@ -0,0 +1,72 @@
+using namespace llvm::util;
+
+namespace {
+
```
Missing test for all the error cases with messages

asudarsa replied:
will address in one of upcoming commits. Thanks

asudarsa replied:
Addressed in 025503af2ec438c066093279210c271f38dc9cc1

Thanks

arsenm commented on llvm/lib/Support/PropertySetIO.cpp:
```diff
@@ -0,0 +1,217 @@
+  copy(P);
+
+  if (P.getType() == BYTE_ARRAY) {
+    auto &ByteArrayVal = std::get<byte *>(P.Val);
```
```suggestion
    auto *ByteArrayVal = std::get<byte *>(P.Val);
```

asudarsa replied:
We need the reference here as we want to set it to nullptr.

asudarsa replied:
Modified code to access P.Val directly. 

Thanks

asudarsa committed:
More changes due to shift from Union to std::variant

Signed-off-by: Arvind Sudarsanam <arvind.sudarsanam@intel.com>

asudarsa commented on llvm/include/llvm/Support/PropertySetIO.h:
```diff
@@ -0,0 +1,259 @@
+  SizeTy size() const {
+    switch (Ty) {
+    case UINT32:
+      return sizeof(std::get<uint32_t>(Val));
```
Yes. Of course. Thanks for the catch. Will update in next commit.

asudarsa replied:
Addressed in 0b4ad716189c3fcfb6e77d299d12c11a3784af0b

Thanks

asudarsa commented on llvm/lib/Support/Base64.cpp:
```diff
@@ -90,3 +90,180 @@ llvm::Error llvm::decodeBase64(llvm::StringRef Input,
+  static inline int decode(char Ch) {
+    if (Ch >= 'A' && Ch <= 'Z') // 0..25
+      return Ch - 'A';
+    else if (Ch >= 'a' && Ch <= 'z') // 26..51
```
Thanks. Will update

asudarsa replied:
Addressed in 0b4ad716189c3fcfb6e77d299d12c11a3784af0b

Thanks

asudarsa commented on llvm/lib/Support/Base64.cpp:
```diff
@@ -90,3 +90,180 @@ llvm::Error llvm::decodeBase64(llvm::StringRef Input,
+
+      if (!TrRes)
+        return TrRes.takeError();
+      // each quad of chars produces three bytes of output
```
Thanks for the suggestion. Checked all comments and updated. Will appear in next commit.

asudarsa replied:
Addressed in https://github.com/llvm/llvm-project/commit/0b4ad716189c3fcfb6e77d299d12c11a3784af0b

Thanks

asudarsa commented on llvm/include/llvm/Support/PropertySetIO.h:
```diff
@@ -0,0 +1,259 @@
+  ~PropertyValue() {
+    if (std::holds_alternative<byte *>(Val)) {
+      byte *ByteArrayVal = std::get<byte *>(Val);
+      delete ByteArrayVal;
```
This is the first time I am using std::variant. I looked up some documentation and this was one of the suggested ways to 'free' memory. I am not sure how to use unique_ptr here. 
Thanks

asudarsa replied:
We have used unique_ptr inside the 'read' routine. That's the only place where we allocate memory for 'Val'. So, by transferring the ownership back to caller from 'read' routine, the caller now has full ownership of the memory allocated to 'Val'.
Hope that's agreeable.
Thanks

asudarsa commented on llvm/lib/Support/PropertySetIO.cpp:
```diff
@@ -0,0 +1,217 @@
+namespace llvm {
+// output a property to a stream
+raw_ostream &operator<<(raw_ostream &Out, const PropertyValue &Prop) {
+  Out << static_cast<int>(Prop.getType()) << "|";
```
Adding this to upcoming commit.

asudarsa replied:
Addressed in https://github.com/llvm/llvm-project/commit/0b4ad716189c3fcfb6e77d299d12c11a3784af0b

Thanks

asudarsa commented on llvm/lib/Support/PropertySetIO.cpp:
```diff
@@ -0,0 +1,217 @@
+} // namespace llvm
+
+void PropertySetRegistry::write(raw_ostream &Out) const {
+  for (const auto &PropSet : PropSetMap) {
+    Out << "[" << PropSet.first << "]\n";
+
+    for (const auto &Props : PropSet.second) {
+      Out << Props.first << "=" << Props.second << "\n";
+    }
```
Adding this to upcoming commit.

asudarsa replied:
Addressed in https://github.com/llvm/llvm-project/commit/0b4ad716189c3fcfb6e77d299d12c11a3784af0b

Thanks

asudarsa commented on llvm/lib/Support/PropertySetIO.cpp:
```diff
@@ -0,0 +1,217 @@
+  constexpr size_t SizeFieldSize = sizeof(SizeTy);
+
+  // Allocate space for size and data.
+  Val = new byte[SizeFieldSize + DataSize];
```
I will give this a try. Thanks

asudarsa commented on llvm/include/llvm/Support/PropertySetIO.h:
```diff
@@ -0,0 +1,259 @@
+  const byte *asRawByteArray() const {
+    if (Ty != BYTE_ARRAY)
+      llvm_unreachable("must be BYTE_ARRAY value");
+    auto &ByteArrayVal = std::get<byte *>(Val);
```
Adding this to upcoming commit. Thanks

asudarsa replied:
Addressed in https://github.com/llvm/llvm-project/commit/0b4ad716189c3fcfb6e77d299d12c11a3784af0b

Thanks

asudarsa commented on llvm/include/llvm/Support/PropertySetIO.h:
```diff
@@ -0,0 +1,259 @@
+//
+//===----------------------------------------------------------------------===//
+// Models a sequence of property sets and their input and output operations.
+// TODO use Yaml as I/O engine.
```
Will remove the TODO. Thanks.

asudarsa commented on llvm/include/llvm/Support/PropertySetIO.h:
```diff
@@ -0,0 +1,259 @@
+  template <typename C, typename T = typename C::value_type>
+  PropertyValue(const C &Data)
+      : PropertyValue(reinterpret_cast<const byte *>(Data.data()),
+                      Data.size() * sizeof(T) * /* bits in one byte */ 8) {}
```
Agreed. Cleaner.

asudarsa replied:
Addressed in 0b4ad716189c3fcfb6e77d299d12c11a3784af0b

Thanks

asudarsa committed:
Address review comments

Signed-off-by: Arvind Sudarsanam <arvind.sudarsanam@intel.com>

asudarsa commented:
Hi @arsenm 

Thanks very much for your time and feedback. I have addressed all your feedback in the latest commit, except for two.
1. Renaming with SYCL prefix
2. Adding tests to check for error messages

I will work on these and post a new commit soon.

Thanks again
Sincerely

asudarsa committed:
Fix clang format issue

Signed-off-by: Arvind Sudarsanam <arvind.sudarsanam@intel.com>

tahonermann commented on llvm/include/llvm/Support/Base64.h:
```diff
@@ -56,6 +56,32 @@ template <class InputBytes> std::string encodeBase64(InputBytes const &Bytes) {
 
 llvm::Error decodeBase64(llvm::StringRef Input, std::vector<char> &Output);
 
+// General-purpose Base64 encoder/decoder class wrapper.
+class Base64 {
+public:
+  using byte = std::byte;
+
+  // Get the size of the encoded byte sequence of given size.
+  static size_t getEncodedSize(size_t SrcSize);
+
+  // Encode a byte sequence of given size into an output stream.
+  // Returns the number of bytes in the encoded result.
+  static size_t encode(const byte *Src, raw_ostream &Out, size_t SrcSize);
+
+  // Get the size of the encoded byte sequence of given size.
+  static size_t getDecodedSize(size_t SrcSize);
+
+  // Decode a sequence of given size into a pre-allocated memory.
+  // Returns the number of bytes in the decoded result or 0 in case of error.
+  static Expected<size_t> decode(const char *Src, byte *Dst, size_t SrcSize);
+
+  // Allocate minimum required amount of memory and decode a sequence of given
+  // size into it.
+  // Returns the decoded result. The size can be obtained via getDecodedSize.
+  static Expected<std::unique_ptr<byte[]>> decode(const char *Src,
+                                                  size_t SrcSize);
+};
+
```
This wrapper class doesn't seem necessary and doesn't mesh well with the Base functions that are already present.

asudarsa replied:
Removed in de198063cd419f1eb3a1fea18226bb5b25272139

Thanks

tahonermann commented on llvm/lib/Support/Base64.cpp:
```diff
@@ -90,3 +90,180 @@ llvm::Error llvm::decodeBase64(llvm::StringRef Input,
+Expected<std::unique_ptr<byte[]>> Base64::decode(const char *Src,
+                                                 size_t SrcSize) {
+  return Base64Impl::decode(Src, SrcSize);
+}
```
LLVM already has a base64 encoder and decoder. What is the motivation for adding another one? This doesn't seem necessary.

asudarsa replied:
Removed in de198063cd419f1eb3a1fea18226bb5b25272139

Thanks

tahonermann commented on llvm/lib/Support/Base64.cpp:
```diff
@@ -90,3 +90,180 @@ llvm::Error llvm::decodeBase64(llvm::StringRef Input,
+      return Res.takeError();
+    return Expected<std::unique_ptr<byte[]>>(std::move(Dst));
+  }
+};
```
While there are occasionally use cases for classes that have all static member functions, this doesn't look like one of them. What is the reason for indirecting the functions below through such an implementation class?

asudarsa replied:
Removed in de198063cd419f1eb3a1fea18226bb5b25272139

Thanks

tahonermann commented on llvm/lib/Support/Base64.cpp:
```diff
@@ -90,3 +90,180 @@ llvm::Error llvm::decodeBase64(llvm::StringRef Input,
+    return DstOff;
+  }
+
+  static Expected<std::unique_ptr<byte[]>> decode(const char *Src,
+                                                  size_t SrcSize) {
+    size_t DstSize = getDecodedSize(SrcSize);
+    std::unique_ptr<byte[]> Dst(new byte[DstSize]);
+    Expected<size_t> Res = decode(Src, Dst.get(), SrcSize);
+    if (!Res)
+      return Res.takeError();
+    return Expected<std::unique_ptr<byte[]>>(std::move(Dst));
+  }
```
The only functions that should be defined inside the class body are ones that are intended to be inline functions. In this case, these functions aren't even defined in a header file; there is no reason for them to be defined inline.

asudarsa replied:
Removed in de198063cd419f1eb3a1fea18226bb5b25272139

Thanks

tahonermann reviewed (changes requested):
I don't see motivation for adding an additional base64 implementation to LLVM. If improvements to the existing base64 implementation are desired or warranted, please evolve the existing implementation rather than adding an additional one.

asudarsa commented:
Hi @tahonermann 

Thanks for the feedback. There is some history behind the two different implementations. One was submitted directly upstream and the other one was developed downstream. 
I do agree with your suggestion. Let me try to use the implementation available upstream and see if it will work for us.

Thanks

asudarsa committed:
Remove changes to Base64 implementation and use existing implementation

Signed-off-by: Arvind Sudarsanam <arvind.sudarsanam@intel.com>

asudarsa committed:
Add SYCL prefix to PropertySet etc.

Signed-off-by: Arvind Sudarsanam <arvind.sudarsanam@intel.com>

asudarsa committed:
Add tests expected to fail and make sure errors are caught

Signed-off-by: Arvind Sudarsanam <arvind.sudarsanam@intel.com>

asudarsa commented:
Hi @arsenm, @jhuber6 and @tahonermann 

Thanks so much for the feedback.
I have added three separate commits to address the concerns here.
Commit 1:
Remove changes to Base64 implementation and use existing implementation
Commit 2:
Add SYCL prefix to PropertySet etc.
Commit 3:
Add tests expected to fail and make sure errors are caught

Please take a look when convenient.

Thanks again
Sincerely

jhuber6 commented on llvm/include/llvm/Support/SYCLPropertySetIO.h:
```diff
@@ -0,0 +1,274 @@
+
+  // Get property value as unsigned 32-bit integer
+  uint32_t asUint32() const {
+    if (Ty != UINT32)
+      llvm_unreachable("must be UINT32 value");
```
Is this an assert with extra steps?

asudarsa replied:
Ah. I looked into llvm_unreachable. yes. We can replace this with an 'assert'. Will add this to upcoming commit. Thanks

asudarsa replied:
Addressed in 4de315891600364fe4798361e3177625afc542b8

Thanks

jhuber6 reviewed (commented):
Seems better, so this is basically a textual format like YAML that this class just parses / decodes? I would've suggested to just use the existing YAML support but I'm guessing this is deep in your toolchain which makes that option very difficult.

asudarsa commented on llvm/lib/Support/Base64.cpp:
```diff
@@ -90,3 +90,180 @@ llvm::Error llvm::decodeBase64(llvm::StringRef Input,
+Expected<std::unique_ptr<byte[]>> Base64::decode(const char *Src,
+                                                 size_t SrcSize) {
+  return Base64Impl::decode(Src, SrcSize);
+}
```
Addressed in 0b4ad716189c3fcfb6e77d299d12c11a3784af0b 
Thanks

asudarsa commented on llvm/lib/Support/PropertySetIO.cpp:
```diff
@@ -0,0 +1,217 @@
+
+PropertyValue::PropertyValue(const byte *Data, SizeTy DataBitSize) {
+  constexpr int ByteSizeInBits = 8;
+  Ty = BYTE_ARRAY;
```
Addressed in https://github.com/llvm/llvm-project/commit/0b4ad716189c3fcfb6e77d299d12c11a3784af0b

Thanks

asudarsa commented on llvm/include/llvm/Support/PropertySetIO.h:
```diff
@@ -0,0 +1,259 @@
+    auto &PropSet = PropSetMap[Category];
+
+    for (const auto &Prop : Props)
+      PropSet.insert_or_assign(Prop.first, PropertyValue(Prop.second));
```
Addressed in https://github.com/llvm/llvm-project/commit/0b4ad716189c3fcfb6e77d299d12c11a3784af0b

Thanks

asudarsa commented on llvm/include/llvm/Support/PropertySetIO.h:
```diff
@@ -0,0 +1,259 @@
+/// content.
+///
+/// The order of keys is preserved and corresponds to the order of insertion.
+class PropertySetRegistry {
```
Addressed in 20a7d75fa6d3c91ea9fdf5638b898e65c76eeefa

Thanks

asudarsa commented on llvm/unittests/Support/PropertySetIOTest.cpp:
```diff
@@ -0,0 +1,72 @@
+    PropSetsPtr->get()->write(OS);
+  }
+  // Check that the original and the serialized version are equal
+  ASSERT_EQ(Serialized, Content);
```
Addressed in https://github.com/llvm/llvm-project/commit/0b4ad716189c3fcfb6e77d299d12c11a3784af0b

Thanks

asudarsa commented:
> I don't see motivation for adding an additional base64 implementation to LLVM. If improvements to the existing base64 implementation are desired or warranted, please evolve the existing implementation rather than adding an additional one.

Changes to Base64 removed here: de198063cd419f1eb3a1fea18226bb5b25272139 Thanks

asudarsa committed:
Update test to pass when assertions are turned on

Signed-off-by: Arvind Sudarsanam <arvind.sudarsanam@intel.com>

asudarsa committed:
Replace simple llvm_unreachable with asserts

Signed-off-by: Arvind Sudarsanam <arvind.sudarsanam@intel.com>

arsenm commented on llvm/include/llvm/Support/SYCLPropertySetIO.h:
```diff
@@ -0,0 +1,268 @@
+    return static_cast<Type>(T);
+  }
+
+  ~SYCLPropertyValue() {}
```
Leaking the byte array case. Just use unique_ptr?

asudarsa replied:
Let me take another look. Thanks

arsenm commented on llvm/include/llvm/Support/SYCLPropertySetIO.h:
```diff
@@ -0,0 +1,268 @@
+  SYCLPropertyValue(const llvm::StringRef &Str)
+      : SYCLPropertyValue(reinterpret_cast<const std::byte *>(Str.data()),
+                          Str.size() * sizeof(char) *
+                              /* bits in one byte */ 8) {}
```
CHAR_BIT

asudarsa replied:
I thought I changed that..sorry...

arsenm commented on llvm/include/llvm/Support/SYCLPropertySetIO.h:
```diff
@@ -0,0 +1,268 @@
+//=- SYCLPropertySetIO.h -- models a sequence of property sets and their I/O =//
```
```suggestion
//=- SYCLPropertySetIO.h -- models a sequence of property sets ----*- C++ -*-=//
```

llvm-beanz commented:
Is there any documentation about how this is used in the SYCL toolchain? It's a bit odd (although not unprecedented) to put toolchain or target specific things into LLVMSupport as opposed to the BinaryFormat or MC libraries.

llvm-beanz commented on llvm/include/llvm/Support/SYCLPropertySetIO.h:
```diff
@@ -0,0 +1,268 @@
+  using SizeTy = uint64_t;
+
+  // Defines supported property types
+  enum Type { first = 0, NONE = first, UINT32, BYTE_ARRAY, last = BYTE_ARRAY };
```
Is there a reason these need to be all-caps or can we make this an enum class and follow LLVM naming conventions?

https://llvm.org/docs/CodingStandards.html#name-types-functions-variables-and-enumerators-properly

llvm-beanz commented on llvm/include/llvm/Support/SYCLPropertySetIO.h:
```diff
@@ -0,0 +1,268 @@
+  void copy(const SYCLPropertyValue &P);
+
+  Type Ty = NONE;
+  std::variant<uint32_t, std::byte *> Val;
```
It's a little redundant to store the type and a variant (which itself stores the type). Should this just be a union?

asudarsa replied:
We initiallly used an 'union' and are trying to 'modernize' the code by using variants. I agree that it is redundant t store the type when using a variant.
Will try to resolve this soon.

Thanks

llvm-beanz commented on llvm/include/llvm/Support/SYCLPropertySetIO.h:
```diff
@@ -0,0 +1,268 @@
+  static constexpr char SYCL_DEVICE_GLOBALS[] = "SYCL/device globals";
+  static constexpr char SYCL_DEVICE_REQUIREMENTS[] = "SYCL/device requirements";
+  static constexpr char SYCL_HOST_PIPES[] = "SYCL/host pipes";
+  static constexpr char SYCL_VIRTUAL_FUNCTIONS[] = "SYCL/virtual functions";
```
These are very much not in the normal LLVM naming convention:
https://llvm.org/docs/CodingStandards.html#name-types-functions-variables-and-enumerators-properly

llvm-beanz reviewed (commented):
Mostly a bunch of stylistic comments. I'm not sure I understand how this code is used or where it needs to fit, but there's a lot of coding standards violations.

asudarsa commented:
> Is there any documentation about how this is used in the SYCL toolchain? It's a bit odd (although not unprecedented) to put toolchain or target specific things into LLVMSupport as opposed to the BinaryFormat or MC libraries.

Hi @llvm-beanz 

Thanks so very for your valuable time and feedback. I will try to address all your comments in an upcoming commit to my best ability. As far as documentation goes, we are working to upstream a high-level SYCL documentation where use of PropertySetIO will be outlined. In the meanwhile, I have added some description inside the summary of this PR. Hope that answers some of your queries here.

Thanks

asudarsa committed:
Addressed the following comments:

1. LLVM standard coding violations pointed out.
2. Removed 'Ty' field in the SYCLPropertyValue class and adjusted the code.

Signed-off-by: Arvind Sudarsanam <arvind.sudarsanam@intel.com>

asudarsa commented on llvm/lib/Support/SYCLPropertySetIO.cpp:
```diff
@@ -0,0 +1,207 @@
+      break;
+    }
+    case SYCLPropertyValue::Type::BYTE_ARRAY: {
+      std::vector<char> Output;
```
Pre-computed the size and used it to resize the Output. However, I continue to use std::vector as decodeBase64 routine in Base64 implementation requires a std::string

Thanks

asudarsa commented on llvm/lib/Support/SYCLPropertySetIO.cpp:
```diff
@@ -0,0 +1,207 @@
+    // See if this line starts a new property set
+    if (LI->starts_with("[")) {
+      // Parse the category (property name)
+      auto EndPos = LI->rfind(']');
```
I have updated most of the sites. 
Thanks for pointing this out.
I am traditionally an 'almost always auto' implementer. But I have adjusted my approach here.

Thanks

asudarsa commented on llvm/lib/Support/SYCLPropertySetIO.cpp:
```diff
@@ -0,0 +1,207 @@
+using namespace llvm::util;
+using namespace llvm;
+
+namespace {
+
+::llvm::Error makeError(const Twine &Msg) {
+  return createStringError(std::error_code{}, Msg);
+}
+
+} // anonymous namespace
```
Updated. Thanks

asudarsa committed:
Avoid memory leak for byte array

Signed-off-by: Arvind Sudarsanam <arvind.sudarsanam@intel.com>

arsenm commented on llvm/include/llvm/Support/SYCLPropertySetIO.h:
```diff
@@ -66,8 +66,13 @@ class SYCLPropertyValue {
+    if (std::holds_alternative<std::byte *>(Val)) {
+      auto ByteArrayVal = std::get<std::byte *>(Val);
+      if (ByteArrayVal)
+        delete[] ByteArrayVal;
```
Please just use unique_ptr

asudarsa replied:
Hi @arsenm 

I will give it another shot. I had a bit of an issue in getting unique_ptr to work with remainder of the code. I must be missing something. I will put up a change soon. Thanks again for looking into this.

Sincerely

asudarsa replied:
Hi @arsenm 

I added use of unique_ptr in 0d5af17a958984fba156a95e6adf7a5b6879c1f2
It was a good learning curve for me.

Thanks
Sincerely

llvm-beanz commented:
> In the meanwhile, I have added some description inside the summary of this PR. Hope that answers some of your queries here.

The description in the summary doesn't really explain how this is used other than "to pass information about specialization constants, etc, to the SYCL runtime."

That's not really descriptive, and doesn't explain why this new textual format was designed. Some specific questions:
1) Is this text format part of the SYCL specification, or is it an implementation detail?
2) If it is an implementation detail, can it be changed to a standard defined textual format (YAML, JSON, XML)? I ask this specifically because custom serialization formats are often great targets for security vulnerabilities, so using a standardized and defined format is preferable to making your own. Obviously if this is a required part of the SYCL specification that has different considerations.
3) How is this used during the compiler? Specifically, where are they generated, and where are they consumed?

llvm-beanz commented on llvm/include/llvm/Support/SYCLPropertySetIO.h:
```diff
@@ -0,0 +1,285 @@
+    else if (Ty == ByteArray)
+      Val = (std::byte *)(0);
+    else
+      llvm_unreachable_internal("unsupported SYCL property type");
```
Why not just `assert(Ty != None)` at the top of the function? Also if this needs to handle all possible cases of `Ty` a switch would be preferable so that you get diagnostics for missing cases.

asudarsa replied:
Used switch-case instead of if-then-else. Thanks

asudarsa committed:
Use unique_ptr for Byte Array

Signed-off-by: Arvind Sudarsanam <arvind.sudarsanam@intel.com>

jhuber6 commented:
> 2\. If it is an implementation detail, can it be changed to a standard defined textual format (YAML, JSON, XML)? I ask this specifically because custom serialization formats are often great targets for security vulnerabilities, so using a standardized and defined format is preferable to making your own. Obviously if this is a required part of the SYCL specification that has different considerations.

My gut instinct is that this should probably be a binary format w/ all the YAML serialization all the others get. But my assumption is that SYCL's implementation uses this somehow so I'm unsure how much work it is to rip it all up and do something else.

asudarsa committed:
Minor change to use switch-case instead of if-then-else

Signed-off-by: Arvind Sudarsanam <arvind.sudarsanam@intel.com>

asudarsa commented:
Hi @llvm-beanz 

Thanks so much for guiding this code review process. I have tried to answer your questions here.

> 1. Is this text format part of the SYCL specification, or is it an implementation detail?
This text format is not part of the SYCL specification. It is used to represent SYCL properties which is part of the SYCL specification, but the format itself is more of an implementation detail.

> 2. If it is an implementation detail, can it be changed to a standard defined textual format (YAML, JSON, XML)? I ask this specifically because custom serialization formats are often great targets for security vulnerabilities, so using a standardized and defined format is preferable to making your own. Obviously if this is a required part of the SYCL specification that has different considerations.
One of my colleagues had tried to switch to YAML some time ago, but that effort got stuck pretty quickly. The main problem that we encountered with switching to YAML was lack of flexibility: we don't know how many SYCL properties there will be and SYCL properties within a single set (which could have been represented as an array) may have different types (which makes it hard to represent them as an array). 

> 3. How is this used during the compiler? Specifically, where are they generated, and where are they consumed?

Here is a concise sequence of steps done during SYCL device code linking.
```
clang: .cpp -> .bc
llvm-link: .bc (multiple) -> .bc (single)
sycl-post-link: .bc (single) -> .bc + .prop (**Generation of data in PropertySetIO format**) 
device-compiler/spirv-llvm-translator here to transform device code: .bc -> .image
clang-offload-wrapper: .image + .prop -> .bc (later linked with host code) (**Consumption of data in PropertySetIO format**) 
```

Hope this gives a better picture of why we need SYCLPropertySetIO and how it is used.

Thanks
Sincerely

asudarsa committed:
Fix formatting issue

Signed-off-by: Arvind Sudarsanam <arvind.sudarsanam@intel.com>

jhuber6 commented:
> sycl-post-link: .bc (single) -> .bc + .prop (**Generation of data in PropertySetIO format**) 
> device-compiler/spirv-llvm-translator here to transform device code: .bc -> .image

Ideally this will be a single step once the backend target is done.

> clang-offload-wrapper: .image + .prop -> .bc (later linked with host code) (**Consumption of data in PropertySetIO format**)

I think I deleted this in community LLVM ages ago, was planning on adding it back in as an llvm tool but never got around to doing it.

asudarsa commented:
> > sycl-post-link: .bc (single) -> .bc + .prop (**Generation of data in PropertySetIO format**)
> > device-compiler/spirv-llvm-translator here to transform device code: .bc -> .image
> 
> Ideally this will be a single step once the backend target is done.
> 
> > clang-offload-wrapper: .image + .prop -> .bc (later linked with host code) (**Consumption of data in PropertySetIO format**)
> 
> I think I deleted this in community LLVM ages ago, was planning on adding it back in as an llvm tool but never got around to doing it.

Yes. That's the objective. We will try to minimize number of steps once the SPIR-V backend is ready. We also do not intend to use 'clang-offload-wrapper' tool. Wrapping will be done using API calls (as is done in the community). 

Thanks @jhuber6

llvm-beanz commented:
>    One of my colleagues had tried to switch to YAML some time ago, but that effort got stuck pretty quickly. The main problem that we encountered with switching to YAML was lack of flexibility: we don't know how many SYCL properties there will be and SYCL properties within a single set (which could have been represented as an array) may have different types (which makes it hard to represent them as an array).

YAML files can contain multiple "documents" as well. I'm not sure I see how YAML is unsuitable for this use case.

That aside, it really seems to me that this design decision has a bunch of different implications.

If we go forward with this solution I think we need _a lot_ more tests of the parser itself since parsing is generally a source of security bugs.

asudarsa commented:
> > One of my colleagues had tried to switch to YAML some time ago, but that effort got stuck pretty quickly. The main problem that we encountered with switching to YAML was lack of flexibility: we don't know how many SYCL properties there will be and SYCL properties within a single set (which could have been represented as an array) may have different types (which makes it hard to represent them as an array).
> 
> YAML files can contain multiple "documents" as well. I'm not sure I see how YAML is unsuitable for this use case.
> 
> That aside, it really seems to me that this design decision has a bunch of different implications.
> 
> If we go forward with this solution I think we need _a lot_ more tests of the parser itself since parsing is generally a source of security bugs.

Hi @llvm-beanz 

Thanks for your feedback. Just a heads up. I am currently looking at adding more tests.
A quick addon comment. One of our primary objective is to provide a fully functional SYCL offloading support and PropertySetIO plays a small, but important role in our compilation flow. So, it will be great to have this upstreamed to unblock other parts of our upstreaming process. One good source of confidence for our implementation is that it has been present in our intel/llvm compiler for a while. However, I completely agree with your pointer that this will benefit from additional testing.

Thanks

asudarsa committed:
Add more tests for PropertySetIO usage

Signed-off-by: Arvind Sudarsanam <arvind.sudarsanam@intel.com>

arsenm commented on llvm/unittests/Support/SYCLPropertySetIOTest.cpp:
```diff
@@ -135,4 +188,127 @@ TEST(SYCLPropertySet, ByteArrayValuesIO) {
+    // Serialize
+    PropSet.write(OS);
+  }
+  llvm::errs() << Serialized << "\n";
+  EXPECT_EQ(Serialized, Expected);
```
```suggestion
  EXPECT_EQ(Serialized, Expected) << Serialized;
```

Same throughout

asudarsa replied:
These statements were for 'local' testing/debugging purpose. Removed them throughout.
Thanks for the catch.

asudarsa commented:
Hi @llvm-beanz 

First of all, thanks so much for your continued reviews here. 
For adding more testing, I looked at unittest/Support/YAMLIOTest.cpp as a reference and I also looked at the use-cases of PropertySetIO in our downstream SYCL compiler.
I have managed to add several test cases covering the use-cases in the SYCL compiler and also several test cases that check if errors are captured as expected.
Kindly take a look when convenient. Please let me know if there are gaps and I will try to handle them.

Thanks again
Sincerely

asudarsa committed:
Remove redundant print statements

Signed-off-by: Arvind Sudarsanam <arvind.sudarsanam@intel.com>

asudarsa commented:
Hi @tahonermann, I have addressed your concern about 'additional base64 implementation' in de198063cd419f1eb3a1fea18226bb5b25272139. I have completely removed the 'additional' base64 implementation and have refactored the PropertySetIO implementation to use community implementation. It will be great if you can take a look when convenient and verify if I have addressed your concerns adequately. Thanks so much.

asudarsa commented:
Hi @arsenm 

Thanks very much for your feedback thus far. Based on your feedback, I am now using 'unique_ptr' in my implementation. I have also tried to address all other concerns to my best ability. Can you please take another look when convenient and see if I have adequately addressed all concerns?

Sincerely

llvm-beanz commented on llvm/unittests/Support/SYCLPropertySetIOTest.cpp:
```diff
@@ -0,0 +1,309 @@
+  }
+  EXPECT_EQ(Serialized, Expected);
+}
+} // namespace
```
What about something like this?
`[SomeValue][OhNoAnotherValue]\nbleh=oops`

asudarsa commented:
Closing this as we are working on alternate approach to implement SYCL property sets.

Thanks
